superset-frontend/src/explore/ExplorePage.tsx

```diff
@@ -38,12 +45,58 @@ const fetchExploreData = () => {
   })(exploreUrlParams);
 };
 
+const useExploreInitialData = (
+  shouldUseDashboardData: boolean,
+  sliceId: string | null,
+) => {
+  const slice = useSelector<RootState, Slice | null>(({ sliceEntities }) =>
+    isDefined(sliceId) ? sliceEntities?.slices?.[sliceId] : null,
+  );
+  const formData = slice?.form_data;
+  const { id: datasourceId, type: datasourceType } = useSelector<
+    RootState,
+    { id: number | undefined; type: string | undefined }
+  >(({ datasources }) =>
+    formData?.datasource
+      ? pick(datasources[formData.datasource], ['id', 'type'])
+      : { id: undefined, type: undefined },
+  );
+  return useCallback(() => {
+    if (
+      !shouldUseDashboardData ||
+      !isDefined(slice) ||
+      !isDefined(formData) ||
+      !isDefined(datasourceId) ||
+      !isDefined(datasourceType)
+    ) {
+      return fetchExploreData();
+    }
+    return SupersetClient.get({
+      endpoint: `/datasource/get/${datasourceType}/${datasourceId}/`,
```

Comment:
there is a new endpoint for the dataset GET method is `/api/v1/dataset/<datasource PK(a.k.a datasource id without type>`. I added it at [here](https://github.com/apache/superset/pull/20677/files#diff-cc681838b8b8b6cb3d1a5dbf02c118d5f3353db8023f49c55f7a673e879be030)

Reply:
I actually used `GET /v1/dataset` endpoint at first, but then changed it to `GET /datasource/get/`. As far as I know, Explore now supports (or will soon support?) other sources than datasets, such as SQL queries. Tagging @hughhhh for transparency.
Also, the `/v1/explore/` endpoint uses `DatasourceDAO` to get datasource and the response from `/v1/dataset` endpoint is missing some fields compared to datasource from `/v1/explore`, which caused some components in Explore to crash

Reply:
I propose we stick to the legacy `/datasource/get` endpoint for now and create a new `/v1/datasource` API in the future and refactor